### PR TITLE
Feature/20461 endpoint rate limiting

### PR DESCRIPTION
**Pull request by jon-shipley**
# Add rate limiting to the pupil auth app

* introduce redis
* move all session handling to redis (dev)
* store rate limiting caches in redis
* the old version of redis accords with Redis in Azure

## New config vars in `pupil-api`
RATE_LIMIT_THRESHOLD=100
RATE_LIMIT_DURATION=60000
RATE_LIMIT_ENABLED=0
REDIS_HOST=localhost
REDIS_PORT=6379
REDIS_KEY=null

### CODE REVIEW

File: pupil-api/src/config.ts [hunk 1]
```diff
@@ -19,9 +24,14 @@ export default {
...
+  Redis: {
+    Host: process.env.REDIS_HOST || 'localhost',
+    Port: process.env.REDIS_PORT || 6379,
+    Key: process.env.REDIS_KEY,
+    useTLS: getEnvironment() === 'Local-Dev' ? false : true
   },
-  Environment: getEnvironment(),
-  AzureStorageConnectionString: process.env.AZURE_STORAGE_CONNECTION_STRING
+  RateLimit: {
+    Threshold: process.env.RATE_LIMIT_THRESHOLD || 100,
```
Comment (GuyHarwood) [inline pupil-api/src/config.ts]: any particular reason why 100?
i wonder how this will affect travis test runs 🤔

File: pupil-api/src/App.ts [hunk 2]
```diff
@@ -9,6 +9,8 @@ import * as helmet from 'helmet'
 import * as uuidV4 from 'uuid/v4'
 import * as winston from 'winston'
 import * as azure from './azure'
+import { rateLimit } from './helpers/rate-limit'
```
Comment (GeorgeChatzigiannis) [inline pupil-api/src/App.ts]: @jon-shipley where is this file? is it pushed to the remote branch?
Can't seem to find it and from my understanding it's a new one

Comment (jon-shipley) [reply pupil-api/src/App.ts]: Oops! 🤦‍♂️ 

I've added it now.  Thank you!

File: pupil-api/src/config.ts (hunk 1 above)
Comment (jon-shipley) [reply pupil-api/src/config.ts]: Seems reasonable for larger schools

**Comment by GuyHarwood** [comment]
Blocked until i have added the env vars to the library for deployment

File: pupil-api/src/config.ts [hunk 3]
```diff
@@ -19,9 +24,15 @@ export default {
...
+    Port: process.env.REDIS_PORT || 6379,
+    Key: process.env.REDIS_KEY,
+    useTLS: getEnvironment() === 'Local-Dev' ? false : true
   },
-  Environment: getEnvironment(),
-  AzureStorageConnectionString: process.env.AZURE_STORAGE_CONNECTION_STRING
+  RateLimit: {
+    Threshold: parseInt(process.env.RATE_LIMIT_THRESHOLD,10) || 100,
+    Duration: parseInt(process.env.RATE_LIMIT_DURATION, 10) || 1000 * 60, // 1 minute in ms
+    Enabled: process.env.hasOwnProperty('RATE_LIMIT_ENABLED') ? toBool(process.env.RATE_LIMIT_ENABLED) : true
```
[the remark this answers is not in the document]
Comment (jon-shipley) [reply pupil-api/src/config.ts]: Done

**Comment by GuyHarwood** [comment]
@jon-shipley deployment configuration libraries updated.  this is now unblocked

**Pull request merged**